src/mono/browser/runtime/diagnostics/diag-js.ts

```diff
@@ -0,0 +1,148 @@
+import { DiagConnectionBase, downloadBlob, fnClientProvider, IDiagClient, IDiagConnection, IDiagSession, schedule_diagnostic_server_loop, SessionId } from "./common";
+import { PromiseAndController } from "../types/internal";
+import { loaderHelpers } from "./globals";
+import { mono_log_warn } from "./logging";
+import { collectCpuSamples } from "./dotnet-cpu-profiler";
+import { collectPerfCounters } from "./dotnet-counters";
+import { collectGcDump } from "./dotnet-gcdump";
+
+//let diagClient:IDiagClient|undefined = undefined as any;
+//let server:DiagServer = undefined as any;
+
+// configure your application
+// .withEnvironmentVariable("DOTNET_DiagnosticPorts", "download:gcdump")
+// or implement function globalThis.dotnetDiagnosticClient with IDiagClient interface
+
+let nextJsClient:PromiseAndController<IDiagClient>;
+let fromScenarioNameOnce = false;
+
+// Only the last which sent advert is receiving commands for all sessions
+export let serverSession:DiagSession|undefined = undefined;
+
+// singleton wrapping the protocol with the diagnostic server in the Mono VM
+// there could be multiple connection at the same time.
+// DS:advert         ->1
+//                     1<- DC1: command to start tracing session
+// DS:OK, session ID ->1
+// DS:advert         ->2
+// DS:events         ->1
+// DS:events         ->1
+// DS:events         ->1
+// DS:events         ->1
+//                     2<- DC1: command to stop tracing session
+// DS:close          ->1
+
+class DiagSession extends DiagConnectionBase implements IDiagConnection, IDiagSession {
+    public session_id: SessionId = undefined as any;
+    public diagClient?: IDiagClient;
+    public stopDelayedAfterLastMessage:number|undefined = undefined;
+    public resumedRuntime = false;
+
+    constructor (public client_socket:number) {
+        super(client_socket);
+    }
+
+    sendCommand (message: Uint8Array): void {
+        if (!serverSession) {
+            mono_log_warn("no server yet");
+            return;
+        }
+        serverSession.respond(message);
+    }
+
+    async connect_new_client () {
+        this.diagClient = await nextJsClient.promise;
+        cleanup_client();
+        const firstCommand = this.diagClient.commandOnAdvertise();
+        this.respond(firstCommand);
+    }
+
+    // this is message from the diagnostic server, which is Mono VM in this browser
+    send (message:Uint8Array):number {
+        schedule_diagnostic_server_loop();
+        if (advert1.every((v, i) => v === message[i])) {
+            // eslint-disable-next-line @typescript-eslint/no-this-alias
+            serverSession = this;
+            this.connect_new_client();
+        } else if (dotnet_IPC_V1.every((v, i) => v === message[i]) && message[16] == CommandSetId.Server) {
+            if (message[17] == ServerCommandId.OK) {
+                if (message.byteLength === 28) {
+                    const view = message.subarray(20, 28);
```

Comment:
Can you please put here some comment about the logic?

Reply:
refactored it into more meaningful methods